router/usecase/quote.go

```diff
@@ -130,3 +134,63 @@ func (q *quoteImpl) String() string {
 func (q *quoteImpl) GetPriceImpact() osmomath.Dec {
 	return q.PriceImpact
 }
+
+// GetInBaseOutQuoteSpotPrice implements domain.Quote.
+func (q *quoteImpl) GetInBaseOutQuoteSpotPrice() osmomath.Dec {
+	return q.InBaseOutQuoteSpotPrice
+}
+
+// quoteExactAmountOut is a quote wrapper for exact out quotes.
+// Note that only the PrepareResult method is different from the quoteImpl.
+type quoteExactAmountOut struct {
+	*quoteImpl              "json:\"-\""
+	AmountIn                osmomath.Int        "json:\"amount_in\""
+	AmountOut               sdk.Coin            "json:\"amount_out\""
+	Route                   []domain.SplitRoute "json:\"route\""
+	EffectiveFee            osmomath.Dec        "json:\"effective_fee\""
+	PriceImpact             osmomath.Dec        "json:\"price_impact\""
+	InBaseOutQuoteSpotPrice osmomath.Dec        "json:\"in_base_out_quote_spot_price\""
+}
+
+// PrepareResult implements domain.Quote.
+// PrepareResult mutates the quote to prepare
+// it with the data formatted for output to the client.
+// Specifically:
+// It strips away unnecessary fields from each pool in the route.
+// Computes an effective spread factor from all routes.
+//
+// Returns the updated route and the effective spread factor.
+func (q *quoteExactAmountOut) PrepareResult(ctx context.Context, scalingFactor osmomath.Dec) ([]domain.SplitRoute, osmomath.Dec, error) {
+	// Prepare exact out in the quote for inputs inversion
+	if _, _, err := q.quoteImpl.PrepareResult(ctx, scalingFactor); err != nil {
+		return nil, osmomath.Dec{}, err
+	}
+
+	// Assign the inverted values to the quote
+	q.AmountOut = q.quoteImpl.AmountIn
+	q.AmountIn = q.quoteImpl.AmountOut
+	q.Route = q.quoteImpl.Route
+	q.EffectiveFee = q.quoteImpl.EffectiveFee
+	q.PriceImpact = q.quoteImpl.PriceImpact
+	q.InBaseOutQuoteSpotPrice = q.quoteImpl.InBaseOutQuoteSpotPrice
+
+	for i, route := range q.Route {
+		route, ok := route.(*RouteWithOutAmount)
+		if !ok {
+			panic("invalid route type") // should never happen
+		}
```

Comment:
Let's bubble up the error rather than panicking so that a client gets a clearer message if this indeed occurs for one or the other reason

Reply:
Adjusted tor return `types.ErrInvalidRouteType` in such scenario.